File: android/app/src/main/java/com/stak/demo/ui/mystak/Collections.kt
package com.stak.demo.ui.mystak

import com.stak.demo.R
import com.stak.demo.ui.MyStakHoldings

/**
 * The My STAK collection catalogue - what every chip on the Overview
 * (1:3155) carries and what the Collection page (1:3333) serves. AI &
 * Tech is the authored frame verbatim; the other five are the shared
 * demo data both platforms use (Codex parity audit, 2026-09-04), so a
 * tapped chip opens ITS collection instead of the sample one.
 * Mirrors the iOS catalogue beside ios/StakDemo/MyStak/CollectionView.swift.
 */
internal data class CollStock(
	val badge: String, val change: String, val up: Boolean,
	val ticker: String, val company: String, val price: String,
)

/** One collection - glass art (imageRes) OR the chip's category icon (iconRes). */
internal data class StakCollection(
	val id: String,
	val name: String,
	// The authored 1:3155 chip text, kept for the record only - nothing
	// renders it since the holdings store drives every count (Codex audit
	// 2026-09-04); see held() / heldCountLabel().
	val countLabel: String,
	val blurb: String,
	val imageRes: Int? = null,
	val iconRes: Int? = null,
	/**
	 * The collection page's 60 hero (1:3357 authors AI & Tech's glass art at
	 * 60). Every collection draws its own art at that size - the four
	 * category icons are the 4x exports of their 1:3155 cat-icon frames
	 * (user, 2026-09-05: the AI & Tech hero is the look; the 36 chip icon
	 * centred in the frame was not).
	 */
	val heroRes: Int,
	val stocks: List<CollStock>,
)

/** Overview grid order (1:3155): two per row, top-left to bottom-right. */
internal val COLLECTIONS = listOf(
	StakCollection(
		id = "aitech",
		name = "AI & Tech",
		countLabel = "5 stocks",
		blurb = "Your highest-conviction growth and AI names.",
		imageRes = R.drawable.ms_coll_aitech,
		heroRes = R.drawable.ms_coll_aitech,
		stocks = listOf(
			CollStock("N", "▲ 2.4%", true, "NVDA", "NVIDIA", "$122.10"),
			CollStock("A", "▲ 1.2%", true, "AAPL", "Apple", "$229.35"),
			CollStock("M", "▼ 0.4%", false, "MSFT", "Microsoft", "$438.20"),
			CollStock("G", "▲ 0.8%", true, "GOOGL", "Alphabet", "$178.90"),
			CollStock("A", "▲ 2.1%", true, "AMD", "Adv Micro", "$164.30"),
		),
	),
	StakCollection(
		id = "finance",
		name = "Finance",
		countLabel = "3 stocks",
		blurb = "The banks and payment rails that move money.",
		imageRes = R.drawable.ms_coll_finance,
		heroRes = R.drawable.ms_coll_finance,
		stocks = listOf(
			CollStock("J", "▲ 0.6%", true, "JPM", "JPMorgan", "$245.60"),
			CollStock("V", "▲ 0.3%", true, "V", "Visa", "$352.10"),
			CollStock("G", "▼ 0.5%", false, "GS", "Goldman Sachs", "$612.40"),
		),
	),
	StakCollection(
		id = "green",
		name = "Green Energy",
		countLabel = "3 stocks",
		blurb = "Solar, the grid and the utilities going clean.",
		iconRes = R.drawable.ic_cat_green,
		heroRes = R.drawable.ms_coll_green,
		stocks = listOf(
			CollStock("E", "▲ 1.9%", true, "ENPH", "Enphase", "$78.40"),
			CollStock("N", "▲ 0.4%", true, "NEE", "NextEra", "$84.20"),
			CollStock("F", "▼ 1.1%", false, "FSLR", "First Solar", "$228.90"),
		),
	),
	StakCollection(
		id = "realestate",
		name = "Real Estate",
		countLabel = "2 stocks",
		blurb = "Your small hedge: warehouses and rent checks.",
		iconRes = R.drawable.ic_cat_realestate,
		heroRes = R.drawable.ms_coll_realestate,
		stocks = listOf(
			CollStock("P", "▲ 0.2%", true, "PLD", "Prologis", "$118.30"),
			CollStock("O", "▼ 0.3%", false, "O", "Realty Income", "$59.10"),
		),
	),
	StakCollection(
		id = "health",
		name = "Healthcare",
		countLabel = "4 stocks",
		blurb = "Drugmakers and insurers with steady demand.",
		iconRes = R.drawable.ic_cat_health,
		heroRes = R.drawable.ms_coll_health,
		stocks = listOf(
			CollStock("L", "▲ 1.4%", true, "LLY", "Eli Lilly", "$792.50"),
			CollStock("U", "▼ 0.8%", false, "UNH", "UnitedHealth", "$318.70"),
			CollStock("J", "▲ 0.5%", true, "JNJ", "Johnson & Johnson", "$162.40"),
			CollStock("P", "▼ 0.2%", false, "PFE", "Pfizer", "$25.30"),
		),
	),
	StakCollection(
		id = "consumer",
		name = "Consumer",
		countLabel = "2 stocks",
		blurb = "Brands people keep buying, in any market.",
		iconRes = R.drawable.ic_cat_consumer,
		heroRes = R.drawable.ms_coll_consumer,
		stocks = listOf(
			CollStock("C", "▲ 0.7%", true, "COST", "Costco", "$947.20"),
			CollStock("N", "▼ 1.3%", false, "NKE", "Nike", "$72.80"),
		),
	),
)

/** The served collection - an unknown id falls back to the authored AI & Tech frame. */
internal fun collection(id: String): StakCollection =
	if (id == OTHER_ID) otherCollection() ?: COLLECTIONS.first()
	else COLLECTIONS.firstOrNull { it.id == id } ?: COLLECTIONS.first()

internal const val OTHER_ID = "other"

/**
 * The stocks the user holds that no collection catalogues - news saves such
 * as AMZN, MU, PLTR, TSLA or XOM (Codex review, PR #166). Built from the
 * holdings and the news feed's stock facts, so every saved ticker has a tile
 * to open and unsave from; null while nothing uncatalogued is held. The demo
 * persona keeps its authored six-tile grid (its TSLA / SNOW are the authored
 * Best / Worst stand-ins). Mirrors ios StakCollections.other.
 */
internal fun otherCollection(): StakCollection? {
	val catalogued = COLLECTIONS.flatMap { it.stocks }.map { it.ticker }.toSet()
	val extra = MyStakHoldings.tickers.filter { it !in catalogued }.sorted()
	if (extra.isEmpty()) return null
	val stocks = extra.map { t ->
		val known = com.stak.demo.ui.news.NewsArticleFeed.hasStockFacts(t)
		val f = com.stak.demo.ui.news.NewsArticleFeed.stockFacts(t)
		CollStock(
			badge = t.take(1),
			change = if (known) (if (f.up) "▲ " else "▼ ") + f.change.removePrefix("+").removePrefix("-").removeSuffix(" today") else "—",
			up = if (known) f.up else true,
			ticker = t,
			company = if (known) f.shortName else t,
			price = if (known) f.price else "—",
		)
	}
	return StakCollection(
		id = OTHER_ID,
		name = "Other",
		countLabel = "",
		blurb = "Stocks you saved from the news that sit outside the six collections.",
		iconRes = R.drawable.ic_saved_bookmark,
		heroRes = R.drawable.ic_saved_bookmark,
		stocks = stocks,
	)
}

/**
 * The collection's stocks the user actually holds. Codex audit
 * (2026-09-04): MyStakHoldings is the single source of truth - Unsave on
 * a tile's Stock Detail drops the stock here, from the chip count and
 * from the Breakdown at once. Reads the observable set, so a composable
 * calling it recomposes when the store changes.
 * Mirrors ios/StakDemo/MyStak/Collections.swift.
 */
internal fun StakCollection.held(): List<CollStock> = stocks.filter { it.ticker in MyStakHoldings.tickers }

/** "1 stock" / "N stocks" - the chip, the Collection hero and the Breakdown share it. */
internal fun heldCountLabel(n: Int): String = if (n == 1) "1 stock" else "$n stocks"

File: ios/StakDemo/MyStak/Collections.swift
import SwiftUI

/// One stock tile in a collection - the fields of the 1:3375 tile template.
/// Mirrors the android/ ui/mystak CollStock.
struct CollStock: Identifiable {
	let badge: String
	let change: String
	let up: Bool
	let ticker: String
	let company: String
	let price: String
	var id: String { ticker }
}

/// One My STAK collection: what its Overview chip (1:3180) and its
/// Collection page hero (1:3333) serve. `image` is glass art (MsColl*),
/// `icon` a category glyph (IcCat*) - a collection carries one or the other.
struct StakCollection: Identifiable {
	let id: String
	let name: String
	/// The authored "5 stocks" label (1:3180 / 1:3333) - kept as the record
	/// of what was drawn; nothing renders it since the holdings store drives
	/// every count (Codex audit 2026-09-04).
	let count: String
	let blurb: String
	var image: String? = nil
	var icon: String? = nil
	/// The collection page's 60 hero (1:3357 authors AI & Tech's glass art at
	/// 60). Every collection draws its own art at that size - the four
	/// category icons are the 4x exports of their 1:3155 cat-icon frames
	/// (user, 2026-09-05). Declared before `stocks`: memberwise-init order.
	let hero: String
	let stocks: [CollStock]

	/// Codex audit (2026-09-04): the stocks of this collection the user
	/// still holds - the chip count, the hero count and the tile grid all
	/// read this, so a deck save or an Unsave moves every number at once.
	/// Mirrors android ui/mystak/Collections.kt.
	func held(in holdings: Set<String>) -> [CollStock] {
		stocks.filter { holdings.contains($0.ticker) }
	}
}

/// "1 stock" / "N stocks" - the chip, hero and allocation count text.
/// Mirrors android ui/mystak/Collections.kt heldCountLabel.
func heldCountLabel(_ n: Int) -> String {
	n == 1 ? "1 stock" : "\(n) stocks"
}

/// Codex parity audit (2026-09-04): every chip carries its own collection
/// and the Collection page serves the tapped one through the same authored
/// template (1:3333) - exactly as the Discover deck's "Learn more" serves
/// the tapped stock. AI & Tech keeps the authored sample verbatim; the
/// other five are the shared demo seed the backend would serve.
/// Mirrors the android/ ui/mystak collection catalogue.
enum StakCollections {
	static let all: [StakCollection] = [
		StakCollection(
			id: "aitech", name: "AI & Tech", count: "5 stocks",
			blurb: "Your highest-conviction growth and AI names.",
			image: "MsCollAITech",
			hero: "MsCollAITech",
			stocks: [
				CollStock(badge: "N", change: "▲ 2.4%", up: true, ticker: "NVDA", company: "NVIDIA", price: "$122.10"),
				CollStock(badge: "A", change: "▲ 1.2%", up: true, ticker: "AAPL", company: "Apple", price: "$229.35"),
				CollStock(badge: "M", change: "▼ 0.4%", up: false, ticker: "MSFT", company: "Microsoft", price: "$438.20"),
				CollStock(badge: "G", change: "▲ 0.8%", up: true, ticker: "GOOGL", company: "Alphabet", price: "$178.90"),
				CollStock(badge: "A", change: "▲ 2.1%", up: true, ticker: "AMD", company: "Adv Micro", price: "$164.30")
			]
		),
		StakCollection(
			id: "finance", name: "Finance", count: "3 stocks",
			blurb: "The banks and payment rails that move money.",
			image: "MsCollFinance",
			hero: "MsCollFinance",
			stocks: [
				CollStock(badge: "J", change: "▲ 0.6%", up: true, ticker: "JPM", company: "JPMorgan", price: "$245.60"),
				CollStock(badge: "V", change: "▲ 0.3%", up: true, ticker: "V", company: "Visa", price: "$352.10"),
				CollStock(badge: "G", change: "▼ 0.5%", up: false, ticker: "GS", company: "Goldman Sachs", price: "$612.40")
			]
		),
		StakCollection(
			id: "green", name: "Green Energy", count: "3 stocks",
			blurb: "Solar, the grid and the utilities going clean.",
			icon: "IcCatGreen",
			hero: "MsCollGreen",
			stocks: [
				CollStock(badge: "E", change: "▲ 1.9%", up: true, ticker: "ENPH", company: "Enphase", price: "$78.40"),
				CollStock(badge: "N", change: "▲ 0.4%", up: true, ticker: "NEE", company: "NextEra", price: "$84.20"),
				CollStock(badge: "F", change: "▼ 1.1%", up: false, ticker: "FSLR", company: "First Solar", price: "$228.90")
			]
		),
		StakCollection(
			id: "realestate", name: "Real Estate", count: "2 stocks",
			blurb: "Your small hedge: warehouses and rent checks.",
			icon: "IcCatRealEstate",
			hero: "MsCollRealEstate",
			stocks: [
				CollStock(badge: "P", change: "▲ 0.2%", up: true, ticker: "PLD", company: "Prologis", price: "$118.30"),
				CollStock(badge: "O", change: "▼ 0.3%", up: false, ticker: "O", company: "Realty Income", price: "$59.10")
			]
		),
		StakCollection(
			id: "health", name: "Healthcare", count: "4 stocks",
			blurb: "Drugmakers and insurers with steady demand.",
			icon: "IcCatHealth",
			hero: "MsCollHealth",
			stocks: [
				CollStock(badge: "L", change: "▲ 1.4%", up: true, ticker: "LLY", company: "Eli Lilly", price: "$792.50"),
				CollStock(badge: "U", change: "▼ 0.8%", up: false, ticker: "UNH", company: "UnitedHealth", price: "$318.70"),
				CollStock(badge: "J", change: "▲ 0.5%", up: true, ticker: "JNJ", company: "Johnson & Johnson", price: "$162.40"),
				CollStock(badge: "P", change: "▼ 0.2%", up: false, ticker: "PFE", company: "Pfizer", price: "$25.30")
			]
		),
		StakCollection(
			id: "consumer", name: "Consumer", count: "2 stocks",
			blurb: "Brands people keep buying, in any market.",
			icon: "IcCatConsumer",
			hero: "MsCollConsumer",
			stocks: [
				CollStock(badge: "C", change: "▲ 0.7%", up: true, ticker: "COST", company: "Costco", price: "$947.20"),
				CollStock(badge: "N", change: "▼ 1.3%", up: false, ticker: "NKE", company: "Nike", price: "$72.80")
			]
		)
	]

	/// The collection an Overview chip pushed; an unknown id serves the
	/// authored AI & Tech sample, like `stockFacts` falls back to AAPL.
	static func collection(_ id: String) -> StakCollection {
		if id == otherId { return other(holdings: MyStakHoldings.shared.tickers) ?? all[0] }
		return all.first { $0.id == id } ?? all[0]
	}

	static let otherId = "other"

	/// The stocks the user holds that no collection catalogues - news saves such
	/// as AMZN, MU, PLTR, TSLA or XOM (Codex review, PR #167). Built from the
	/// holdings and the news feed's stock facts, so every saved ticker has a tile
	/// to open and unsave from; nil while nothing uncatalogued is held. The demo
	/// persona keeps its authored six-tile grid (its TSLA / SNOW are the authored
	/// Best / Worst stand-ins). Mirrors android otherCollection().
	static func other(holdings: Set<String>) -> StakCollection? {
		let catalogued = Set(all.flatMap { $0.stocks }.map(\.ticker))
		// The persona's seeded TSLA/SNOW are the frame's until the persona saves one itself -
		// a recorded save day makes it a real, reversible save (Codex review, PR #166 mirror).
		let extra = holdings.filter { !catalogued.contains($0) && !(StakStore.demoAccount && MyStakHoldings.seed.contains($0) && MyStakHoldings.shared.daysSinceSaved($0) == nil) }.sorted()
		if extra.isEmpty { return nil }
		let stocks = extra.map { t -> CollStock in
			let known = NewsArticleFeed.hasStockFacts(t)
			let f = NewsArticleFeed.stockFacts(t)
			let move = f.change.replacingOccurrences(of: "+", with: "").replacingOccurrences(of: "-", with: "").replacingOccurrences(of: " today", with: "")
			return CollStock(
				badge: String(t.prefix(1)),
				change: known ? (f.up ? "▲ " : "▼ ") + move : "—",
				up: known ? f.up : true,
				ticker: t,
				company: known ? f.shortName : t,
				price: known ? f.price : "—"
			)
		}
		return StakCollection(
			id: otherId, name: "Other", count: "",
			blurb: "Stocks you saved from the news that sit outside the six collections.",
			icon: "IcSavedBookmark", hero: "IcSavedBookmark", stocks: stocks
		)
	}
}
